### PR TITLE
changefeedccl: fix bug in table history when fetching old table descs after restart

### LINKED ISSUE #41694: changefeedccl: bug in table history wrt fetching old table descriptor versions after a restart
In `poller.go` the poller constructs a `tableHistory` object and periodically calls `ExportRequest` with a `startTS` and `endTS` to fetch a list of table descriptors that *were modified* in this time window. Note that this doesn't fetch the table descriptor versions that were live at `startTS`.

Now, consider the case where a backfill is *detected*, but there is a job restart before the backfill is  *triggered*: the `poller` restarts at the previous globally resolved timestamp and starts calling `ExportRequest` again with `startTS` set to the aforementioned globally resolved timestamp. Now, since this `ExportRequest` call doesn't fetch the table descriptor version *live* at `startTS`, 
it is possible that this method only fetches the table descriptor(s) *after* the column backfill mutation has already been applied. 

If this happens, the `tableHistory` and `poller` have no way of detecting and triggering a backfill, which could lead to certain row updates simply never being emitted.

### PR DESCRIPTION
**Pull request by aayushshah15**
Fixes: #41694

In poller.go the poller constructs a tableHistory object and
periodically calls ExportRequest with a startTS and endTS to
fetch a list of table descriptors that were modified in this
time window. Note that this doesn't fetch the table descriptor
versions that were live at startTS.

Now, consider the case where a backfill is detected, but there is a
job restart before the backfill is triggered: the poller restarts
at the previous globally resolved timestamp and starts calling
ExportRequest again with startTS set to the aforementioned globally
resolved timestamp. Now, since this ExportRequest call doesn't
fetch the table descriptor version live at startTS, it is possible
that this method only fetches the table descriptor(s) after the
column backfill mutation has already been applied.

If this happens, the tableHistory and poller have no way of detecting
and triggering a backfill, which could lead to certain row updates
simply never being emitted. This PR fixes this problem by separately
adding the *current live* table descriptor version to the list returned
by `fetchTableDescriptorVersions`.

Release note (changefeed): Fix bug in tableHistory/poller.

Release justification (changefeed): Fixes correctness bug in tableHistory/poller.

Release note: None

### CODE REVIEW

**Comment by aayushshah15** [comment]
@danhhz this is ready for a look. 

I don't intend for the nemeses related changes to be in with this PR, I'm including them so you can get a rough idea of the testing approach I had in mind. I'm currently in the process of cleaning up/testing the rest of the changes.

**Review by danhhz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41702)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @ajwerner, and @danhhz)

___
*[pkg/ccl/changefeedccl/table_history.go, line 308 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41702#-LrQDY-iE_y76vqiSJE3:-LrQDY-iE_y76vqiSJE4:b-4wdh3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/493384c1556e0f94e6fe89244ca9efbc2b1aeeca/pkg/ccl/changefeedccl/table_history.go#L308)):*
> ```Go
> 		return tableDescs, nil
> 	}
> 	err := db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
> ```

we can do this once inside `makeTableHistory` instead, right?
___
*[pkg/ccl/changefeedccl/cdctest/nemeses.go, line 401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41702#-LrQDxZt6bWIL6H7TNtl:-LrQDxZt6bWIL6H7TNtm:b1bf9pe) ([raw file](https://github.com/cockroachdb/cockroach/blob/493384c1556e0f94e6fe89244ca9efbc2b1aeeca/pkg/ccl/changefeedccl/cdctest/nemeses.go#L401)):*
> ```Go
> }
> 
> func addOrRemoveColumn(a fsm.Args) error {
> ```

when you get around to cleaning up and PRing the nemesis changes, this should be two event types, if possible
___
*[pkg/ccl/changefeedccl/cdctest/nemeses.go, line 404 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41702#-LrQEJjICbRjLIqqrBSS:-LrQEJjICbRjLIqqrBST:bqajjx8) ([raw file](https://github.com/cockroachdb/cockroach/blob/493384c1556e0f94e6fe89244ca9efbc2b1aeeca/pkg/ccl/changefeedccl/cdctest/nemeses.go#L404)):*
> ```Go
> 	ns := a.Extended.(*nemeses)
> 
> 	if ns.txn != nil {
> ```

I think we could avoid this if the txn nilness was added to the state machine state

**Comment by danhhz** [comment]
Am I right that the only non-test, non-debugging change in here is the scan to prime the table history? If so, approach lgtm, though I think we can do it in table history construction which is a lot cleaner

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41702)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @ajwerner, and @danhhz)

___
*[pkg/ccl/changefeedccl/table_history.go, line 308 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41702#-LrQDY-iE_y76vqiSJE3:-LrQJZWQC06SAG4RmTis:b-b0bfu8) ([raw file](https://github.com/cockroachdb/cockroach/blob/493384c1556e0f94e6fe89244ca9efbc2b1aeeca/pkg/ccl/changefeedccl/table_history.go#L308)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

we can do this once inside `makeTableHistory` instead, right?
</blockquote></details>

err, it would change the table history abstraction quite a bit. Currently the table history doesn't know anything about previous versions or anything like that. That being said, I think we could indeed lift this logic to the poller constructor.

**Review by danhhz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41702)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15)

___
*[pkg/ccl/changefeedccl/table_history.go, line 308 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41702#-LrQDY-iE_y76vqiSJE3:-LrQKgJQ9byBGDMBc0V3:b-yysgxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/493384c1556e0f94e6fe89244ca9efbc2b1aeeca/pkg/ccl/changefeedccl/table_history.go#L308)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

err, it would change the table history abstraction quite a bit. Currently the table history doesn't know anything about previous versions or anything like that. That being said, I think we could indeed lift this logic to the poller constructor. 
</blockquote></details>

yeah, that makes sense

**Review by danhhz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41702)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15)

___
*[pkg/ccl/changefeedccl/table_history.go, line 308 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41702#-LrQDY-iE_y76vqiSJE3:-LrQLIdt-oumjCbObMc9:b-gkiwmv) ([raw file](https://github.com/cockroachdb/cockroach/blob/493384c1556e0f94e6fe89244ca9efbc2b1aeeca/pkg/ccl/changefeedccl/table_history.go#L308)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

yeah, that makes sense
</blockquote></details>

in fact, it probably belongs in `RunUsingRangefeeds` which even means we don't have to plumb back an error from makePoller (which is super annoying because of the way the changeAggregator processor works)

**Review by ajwerner — APPROVED** [review]
:lgtm: 

I guess you need to wait for your other commit to merge  to rebase it before bors'ing this one? 

> Reviewed 3 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41702#-:-LrQZyYJ7rGar-0cLh-d:ba6j6kj)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15)

___
*[pkg/ccl/changefeedccl/poller.go, line 113 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41702#-LrQYnBSBcRKWFBkUgjX:-LrQYnBTCoewTpPHvjcd:b-5resb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/23744b90287446e94ead1013a989497ca667b6bf/pkg/ccl/changefeedccl/poller.go#L113)):*
> ```Go
> 	}
> 	p.tableHist = makeTableHistory(p.validateTable, highWater)
> ```

nit: extraneous newline
___
*[pkg/ccl/changefeedccl/poller.go, line 302 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41702#-LrQZugwA5d4znChVpBl:-LrQZugwA5d4znChVpBm:b-6glwr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/23744b90287446e94ead1013a989497ca667b6bf/pkg/ccl/changefeedccl/poller.go#L302)):*
> ```Go
> 					boundaryBreak := false
> 					// Make sure scan boundaries less than or equal to `resolvedTS` were
> 					// added to the `scanBoundaries` list before proceeding.
> ```

This is from the previous commit right?

**Comment by aayushshah15** [comment]
___
*[pkg/ccl/changefeedccl/poller.go, line 302 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41702#-LrQZugwA5d4znChVpBl:-LrQbvA63pRzGPNyRf7r:bv02kum) ([raw file](https://github.com/cockroachdb/cockroach/blob/23744b90287446e94ead1013a989497ca667b6bf/pkg/ccl/changefeedccl/poller.go#L302)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This is from the previous commit right? 
</blockquote></details>

Yup, will rebase once the other thing is b o r s ed.

**Comment by danhhz** [comment]
Looks like the other PR merged. Wanna rebase this? It generally lgtm, modulo it'd be nice to include the test I'm working on for it

**Comment by aayushshah15** [comment]
TFTR!

thanks also for the unit test @danhhz, I applied your patch to the PR. Going to b o r s it once teamcity build is happy.

**Comment by danhhz** [comment]
bors r=danhhz,rjwerner

**Pull request merged**
